pkg/funky/server.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+///////////////////////////////////////////////////////////////////////
+package funky
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Server an interface for managing function servers
+type Server interface {
+	GetPort() uint16
+	Invoke(input Message) (io.ReadCloser, error)
+	Stdout() []string
+	Stderr() []string
+	Start() error
+	Shutdown() error
+	Terminate() error
+}
+
+// DefaultServer a struct to hold information about running servers
+type DefaultServer struct {
+	port   uint16
+	cmd    *exec.Cmd
+	client *http.Client
+
+	lock   sync.RWMutex
+	stdout []string
+	stderr []string
+}
+
+// NewServer returns a new DefaultServer with the given port and command
+func NewServer(port uint16, cmd *exec.Cmd) (*DefaultServer, error) {
+	if port < 1024 {
+		return nil, IllegalArgumentError("port")
+	}
+
+	cmd.Env = append(os.Environ(), fmt.Sprintf("PORT=%d", port))
+
+	return &DefaultServer{
+		port:   port,
+		cmd:    cmd,
+		client: &http.Client{},
+	}, nil
+}
+
+// ServerFactory an interface for creating new Servers decoupled from the concrete implementation
+type ServerFactory interface {
+	CreateServer(port uint16) (Server, error)
+}
+
+// DefaultServerFactory concrete implementation of ServerFactory.
+type DefaultServerFactory struct {
+	cmd  string
+	args []string
+}
+
+// NewDefaultServerFactory a DefaultServerFactory constructor; validates the server command.
+func NewDefaultServerFactory(serverCmd string) (ServerFactory, error) {
+	cmds := strings.Fields(serverCmd)
+
+	if len(cmds) < 1 {
+		return nil, IllegalArgumentError(serverCmd)
+	}
+
+	return &DefaultServerFactory{
+		cmd:  cmds[0],
+		args: cmds[1:],
+	}, nil
+}
+
+// CreateServer creates a new server by initiating a Command with the given port and preconfigured server command
+func (f *DefaultServerFactory) CreateServer(port uint16) (Server, error) {
+	cmd := exec.Command(f.cmd, f.args...)
+	return NewServer(port, cmd)
+}
+
+// GetPort returns the port this server is running on
+func (s *DefaultServer) GetPort() uint16 {
+	return s.port
+}
+
+// Invoke calls the server with the given input to invoke a Dispatch function
+func (s *DefaultServer) Invoke(input Message) (io.ReadCloser, error) {
+	p, err := json.Marshal(input)
+
+	timeout := time.Until(input.Context.Deadline)
+
+	if timeout < 0 {
+		return nil, TimeoutError("Did not invoke, already exceeded timeout")
+	}
+
+	s.client.Timeout = timeout
+
+	s.resetStreams()
+
+	url := fmt.Sprintf("http://127.0.0.1:%d", s.GetPort())
+	resp, err := s.client.Post(url, "application/json", bytes.NewBuffer(p))
+
+	if err != nil {
+		if strings.Contains(err.Error(), "Client.Timeout") {
+			return nil, TimeoutError(timeout)
+		} else if strings.Contains(err.Error(), "connection refused") {
+			return nil, ConnectionRefusedError(url)
+		}
+	}
```

Comment:
What do we do if the `err != nil`, but is neither a timeout or connection refused?